arping.c

```diff
@@ -317,11 +317,22 @@ static int finish(struct run_state *ctl)
 		printf("\n");
 		fflush(stdout);
 	}
+
+	/* arping exit code evaluation */
 	if (ctl->dad)
-		return (!!ctl->received);
+		return !!ctl->received;
+
 	if (ctl->unsolicited)
 		return 0;
-	return (!ctl->received);
+
+	if (ctl->timeout) {
+		if (ctl->count > 0)
+			return !(ctl->count <= ctl->received);
+		else
+			return !(ctl->received > 0);
```

Comment:
nitty, but you use `!(ctl->received > 0)` and `!ctl->received` I guess I would use only one of it everywhere...

and yes, what you mentioned here https://github.com/iputils/iputils/issues/538#issuecomment-2146244343 would help a lot :-)

Reply:
Thanks, correct! You're right, that `ctl->received > 0` can be shortened just to `ctl->received`, which makes the condition simpler (only `-c` together with `-w` counts how many replies we get):

```C
/* arping exit code evaluation */
if (ctl->dad)
	return !!ctl->received;

if (ctl->unsolicited)
	return 0;

if (ctl->timeout && ctl->count > 0)
	return !(ctl->count <= ctl->received);

return !ctl->received;
```

`ctl->count` and `ctl->interval` are by default `-1`, therefore they need to be checked against `0`.

> and yes, what you mentioned here https://github.com/iputils/iputils/issues/538#issuecomment-2146244343 would help a lot :-)

I'll guess I merge for now just this fix (to minimize changes) and send another PR which implements this + create exit code section for arping man page. There are many parts which could be improved, but it's a question how much of it should be done (as for all iputils tools: a complete rewrite would be the best to cleanup).